libraries/AP_Proximity/AP_Proximity_ARS408_CAN.cpp

```diff
@@ -0,0 +1,433 @@
+{
+  
+}
+
+ void AP_Proximity_ARS408_CAN::initialize(void)
+ {
+    if (!_initialized) {
+        // Store in NVM
+        set_store_in_nvm(true);
+
+        // Set output type [NOTE:clusters(0x02);objects(0x01)]
+        set_output_type(0x01);
+
+        // Set output sort
+        set_sort_index(1);
+
+        // Set max distance
+        set_max_distance(_distance_max);
+
+        // Set sensor id
+        set_sensor_id(radar_id.get());
+
+        // Set radar power
+        set_radar_power(0);
+
+        // Set quality output[true]
+        set_send_quality(1);
+
+        // Set ext_info[true]
+        set_send_ext_info(1);
+
+        // Set control relay info
+        set_ctrl_relay_cfg(0);
+
+        AP_HAL::CANFrame frame;
+        frame = {(0x200 & AP_HAL::CANFrame::MaskStdID), radar_cfg_msg.raw_data, sizeof(radar_cfg_msg.raw_data)};
+        multican->write_frame(frame, AP_HAL::native_micros64() + 1000);
+
+        _initialized = true;
+    } else {
+        // Send YWA_RATE and SPEED message
+        send_yaw_rate_message();
+        send_speed_message();
+    }
+ }
+
+// update state
+void AP_Proximity_ARS408_CAN::update(void)
+{
+    initialize();
+
+    if ((_last_reading_ms == 0) || (AP_HAL::millis() - _last_reading_ms > 3000)) {
+        set_status(AP_Proximity::Status::NoData);
+    } else {
+         set_status(AP_Proximity::Status::Good);
+    }  
+}
+
+
+void AP_Proximity_ARS408_CAN::send_speed_message()
+{
+    float speed = 0.0f;
+    if (!get_forward_speed(speed)) {
+        // we expect caller will not try to control heading using rate control without a valid speed estimate
+        // on failure to get speed we do not attempt to steer
+        return;
+    }
+
+    // Set speed
+    uint32_t radar_speed = static_cast<uint32_t>(fabs(speed) / 0.02);
+    speed_information speed_information_msg;
+    speed_information_msg.data.RadarDevice_Speed1 = static_cast<uint64_t>(radar_speed >> 8);
+    speed_information_msg.data.RadarDevice_Speed2 = static_cast<uint64_t>(radar_speed & 255);
+    speed_information_msg.data.RadarDevice_SpeedDirection = RadarDevice_SpeedDirection::STANDSTILL;
+    if(speed > 0.2){speed_information_msg.data.RadarDevice_SpeedDirection  = RadarDevice_SpeedDirection::FORWARD;}
+    if(speed < -0.2){speed_information_msg.data.RadarDevice_SpeedDirection = RadarDevice_SpeedDirection::BACKWARD;}
+
+    // construct AP_HAL::CANframe
+    AP_HAL::CANFrame speed_frame;
+    speed_frame = {(0x300 & AP_HAL::CANFrame::MaskStdID), speed_information_msg.raw_data, sizeof(speed_information_msg.raw_data)};
+    multican->write_frame(speed_frame, AP_HAL::native_micros64() + 1000);
+
+}
+
+// get forward speed in m/s (earth-frame horizontal velocity but only along vehicle x-axis).  returns true on success
+bool AP_Proximity_ARS408_CAN::get_forward_speed(float &speed) const
+{
+    Vector3f velocity;
+    const AP_AHRS &_ahrs = AP::ahrs();
+    if (!_ahrs.get_velocity_NED(velocity)) {
```

Comment:
Consider inverting this check and using an early-return model so the indenting isn't as bad in here